frontend/src/metabase/actions/containers/ActionCreator/ActionCreatorView.tsx

```diff
@@ -0,0 +1,120 @@
+  EditorContainer,
+  ModalRoot,
+  ModalActions,
+  ModalLeft,
+} from "metabase/actions/containers/ActionCreator/ActionCreator.styled";
+
+import type NativeQuery from "metabase-lib/queries/NativeQuery";
+
+import Question from "metabase-lib/Question";
+
+import type { SideView } from "./types";
+
+interface ActionCreatorProps {
+  isNew: boolean;
+  canSave: boolean;
+
+  question: Question;
+
+  onChangeQuestionQuery: (query: NativeQuery) => void;
+  onChangeName: (name: string) => void;
+  onCloseModal?: () => void;
+  onChangeFormSettings: (formSettings: ActionFormSettings) => void;
+  onClickSave: () => void;
+  onClickExample: () => void;
+}
+
+const DEFAULT_SIDE_VIEW: SideView = "actionForm";
+
+export default function ActionCreatorView({
+  isNew,
+  canSave,
+  question,
+  onChangeQuestionQuery,
+  onChangeName,
+  onCloseModal,
+  onChangeFormSettings,
+  onClickSave,
+  onClickExample,
+}: ActionCreatorProps) {
+  const [activeSideView, setActiveSideView] =
+    useState<SideView>(DEFAULT_SIDE_VIEW);
+
+  const toggleDataRef = useCallback(() => {
+    setActiveSideView(activeSideView => {
+      if (activeSideView !== "dataReference") {
+        return "dataReference";
+      }
+
+      return DEFAULT_SIDE_VIEW;
+    });
+  }, []);
+
+  const closeSideView = useCallback(() => {
+    setActiveSideView(DEFAULT_SIDE_VIEW);
+  }, []);
+  return (
+    <Modal wide onClose={onCloseModal}>
+      <ModalRoot>
+        <ActionCreatorBodyContainer>
+          <ModalLeft>
+            <DataReferenceTriggerButton onClick={toggleDataRef} />
+            <ActionCreatorHeader
+              type="query"
+              name={question.displayName() ?? t`New Action`}
+              onChangeName={onChangeName}
+            />
+            <EditorContainer>
+              <QueryActionEditor
+                query={question.query() as NativeQuery}
+                onChangeQuestionQuery={onChangeQuestionQuery}
+              />
+            </EditorContainer>
+            <ModalActions>
+              <Button onClick={onCloseModal} borderless>
+                {t`Cancel`}
+              </Button>
+              <Button primary disabled={canSave} onClick={onClickSave}>
+                {isNew ? t`Save` : t`Update`}
+              </Button>
+            </ModalActions>
+          </ModalLeft>
+
+          {
+            (
+              {
+                dataReference: <DataReferenceInline onClose={closeSideView} />,
+                actionForm: (
+                  <FormCreator
+                    params={question?.parameters() ?? []}
+                    formSettings={
+                      question?.card()
+                        ?.visualization_settings as ActionFormSettings
+                    }
+                    onChange={onChangeFormSettings}
+                    onExampleClick={onClickExample}
+                  />
+                ),
+              } as Record<SideView, React.ReactElement>
+            )[activeSideView]
+          }
```

Comment:
Not sure I've seen this rendering pattern a lot. Should we maybe make it a function?

Reply:
I was reluctant on making this a function primarily because I want this to feel like they're all inline so that we see all the markup in the same place rather than hiding that in a separate function. I do agree that this is a bit unconventional still I think sometimes I feel like I need to jump to a lot of places to understand how the component will render if we have lots of these extracted elements, though there is only 1 in this case.

Let's leave this for now and decided if we want to change it again in https://github.com/metabase/metabase/pull/27809